File: data_structures/stack/stack.js
class StackNode {
  constructor(value) {
    this.value = value;
    this.min = value;
    this.max = value;
    this.next = null;
  }
}

class Stack {
  constructor() {
    this.head = null;
  }

  push(value) {
    let node = new StackNode(value);

    if (this.head == null) {
      this.head = node;
      return;
    }

    node.min = value < this.head.min ? value : this.head.min;
    node.max = value > this.head.max ? value : this.head.max;

    let tmp = this.head;
    node.next = tmp;
    this.head = node;
  }

  pop() {
    if (this.head == null) {
      return null;
    }

    let tmp = this.head;
    this.head = tmp.next;
    return tmp.value;
  }

  min() {
    return this.head ? this.head.min : -1;
  }

  max() {
    return this.head ? this.head.max : -1;
  }
}

var stack = new Stack();
stack.push(3);
stack.push(2);
stack.push(1);
console.log(stack.min());
console.log(stack.max());
console.log(stack.pop());
console.log(stack.min());
console.log(stack.max());

File: data_structures/stack/stack.go
package stack

import (
  "fmt"
)

type stackNode struct {
  data int
  min int
  max int
  next *stackNode
}

type Stack struct {
  head *stackNode
}

func NewStack() *Stack {
  return &Stack{}
}


func (s *Stack) Push(value int) error {
  node := &stackNode{
    data: value,
    min: value,
    max: value,
  }

  if s.head == nil {
    s.head = node
    return nil
  }

  if value > s.head.min {
    node.min = s.head.min
  }

  if value < s.head.max {
    node.max = s.head.max
  }

  tmp := s.head
  node.next = tmp
  s.head = node
  return nil
}

func (s *Stack) Pop() (int, error) {
  if s.head == nil {
    return -1, fmt.Errorf("can't pop from an empty stack")
  }

  tmp := s.head
  s.head = tmp.next
  return tmp.data, nil
}

func (s *Stack) Min() (int, error) {
  if s.head == nil {
    return -1, fmt.Errorf("can't get the min value of an empty stack")
  }

  return s.head.min, nil
}

func (s *Stack) Max() (int, error) {
  if s.head == nil {
    return -1, fmt.Errorf("can't get the max value of an empty stack")
  }

  return s.head.max, nil
}
